src/Protocol/Diaspora/Repository/DiasporaContact.php

```diff
@@ -0,0 +1,244 @@
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace Friendica\Protocol\Diaspora\Repository;
+
+use Friendica\BaseRepository;
+use Friendica\Core\System;
+use Friendica\Database\Database;
+use Friendica\Database\Definition\DbaDefinition;
+use Friendica\Model\APContact;
+use Friendica\Model\Contact;
+use Friendica\Model\Item;
+use Friendica\Model\ItemURI;
+use Friendica\Network\HTTPException;
+use Friendica\Protocol\Diaspora\Entity;
+use Friendica\Protocol\Diaspora\Factory;
+use Friendica\Util\DateTimeFormat;
+use GuzzleHttp\Psr7\Uri;
+use Psr\Http\Message\UriInterface;
+use Psr\Log\LoggerInterface;
+
+class DiasporaContact extends BaseRepository
+{
+	const ALWAYS_UPDATE                 = true;
+	const NEVER_UPDATE                  = false;
+	const UPDATE_IF_MISSING_OR_OUTDATED = null;
+
+	protected static $table_name = 'diaspora-contact-view';
+
+	/** @var Factory\DiasporaContact */
+	protected $factory;
+	/** @var DbaDefinition */
+	private $definition;
+
+	public function __construct(DbaDefinition $definition, Database $database, LoggerInterface $logger, Factory\DiasporaContact $factory)
+	{
+		parent::__construct($database, $logger, $factory);
+
+		$this->definition = $definition;
+	}
+
+	/**
+	 * @param array $condition
+	 * @param array $params
+	 * @return Entity\DiasporaContact
+	 * @throws HTTPException\NotFoundException
+	 */
+	public function selectOne(array $condition, array $params = []): Entity\DiasporaContact
+	{
+		return parent::_selectOne($condition, $params);
+	}
+
+	/**
+	 * @param int $uriId
+	 * @return Entity\DiasporaContact
+	 * @throws HTTPException\NotFoundException
+	 */
+	public function selectOneByUriId(int $uriId): Entity\DiasporaContact
+	{
+		return $this->selectOne(['uri-id' => $uriId]);
+	}
+
+	/**
+	 * @param UriInterface $uri
+	 * @return Entity\DiasporaContact
+	 * @throws HTTPException\NotFoundException
+	 */
+	public function selectOneByUri(UriInterface $uri): Entity\DiasporaContact
+	{
+		try {
+			return $this->selectOne(['`url` = ?', (string) $uri]);
+		} catch (HTTPException\NotFoundException $e) {
+		}
+
+		try {
+			return $this->selectOne(['`addr` = ?', (string) $uri]);
+		} catch (HTTPException\NotFoundException $e) {
+		}
+
+		return $this->selectOne(['`alias` = ?', (string) $uri]);
+	}
+
+	/**
+	 * @param UriInterface $addr
+	 * @return Entity\DiasporaContact
+	 * @throws HTTPException\NotFoundException
+	 */
+	public function selectOneByAddr(UriInterface $addr): Entity\DiasporaContact
```

Comment:
What I don't understand: What is this "UriInterface" and why are we using it? As far as I understood this means that doing so we have to do something similar to casting to make this parameter valid? Doesn't it just adds another level of complexity?

Reply:
I do understand this for `selectOneByUri` - but the handle is not an URI since it is missing basic parts like the schema.

Reply:
Schema is only relevant for UR**L**, but isn't necessary for a UR**I** that I know of. In this sense a URL is a more specific subset of a URI.

Reply:
See https://www.rfc-editor.org/rfc/rfc3986#section-3.1
>  Each URI begins with a scheme name that refers to a specification for
   assigning identifiers within that scheme.  As such, the URI syntax is
   a federated and extensible naming system wherein each scheme's
   specification may further restrict the syntax and semantics of
   identifiers using that scheme.

Reply:
Thanks, I'll create a new specific type for the handle then. I thought Webfinger addresses were URIs, but it looks like the actual WebFinger URI is `acct:user@host` so it has a scheme.

Reply:
Having a specific type sounds reasonable. But we should still keep storing only the `user@host` since we can imply the `acct:` at any time here.